### ISSUE TITLE
Unable to detect magic __get properties correctly registered with __isset

### ISSUE THREAD

**Issue by tcarrio** [ramsey/collection#125]
## Description

This was specifically found when working with the `column(...)` method in the `AbstractCollection` implementation. The issue is that PHP's magic getter functionality is not supported.

## Steps to reproduce

1. Write a class Example that defines no properties with a `__get($key)` and `__isset($key)` method to return `"value"` and `true` when the `$key` is `"key"`, respectively
2. Make a new Collection of Examples, e.g. `new Collection([new Example()])`
3. Call `$collection->column('key')`
4. A `ValueExtractionException` is thrown

## Expected behavior

An array is returned: `["value"]` 

## Environment details

- version of this package: *1.3.0*
- PHP version: *8.1*
- OS: *macOS*

## Additional context

Maybe I can get to a PR on this- but no _current_ capacity for it

**Comment by SimoTod** [comment]
I think you just need to add
```
        if (method_exists($element, '__get')) {
            return $element->$propertyOrMethod;
        }
```
[here](https://github.com/ramsey/collection/blob/main/src/Tool/ValueExtractorTrait.php#L74) to support it.

**Comment by tcarrio** [comment]
@SimoTod I may be wrong, but I understood best practice was to utilize both `__isset` as a way to dictate when `__get` _would_ return a value. This feels safer to me than making all classes with `__get` return a value. I had thrown together a patch for this and could push a PR shortly with testing.

**Comment by SimoTod** [comment]
> @SimoTod I may be wrong, but I understood best practice was to utilize both `__isset` as a way to dictate when `__get` _would_ return a value. This feels safer to me than making all classes with `__get` return a value. I had thrown together a patch for this and could push a PR shortly with testing.

Sure, that's best practice but you need to document it in this case because it's not required. 

You can have a magic __get in a standard php class and that would work when you call an arbitrary non existing property.

All class with __get, would call the magic getter if defined and __get, depending on the implementation, may as well throw an exception if not trying to access something not expected.

Slightly related, in theory, you could make the same reasoning for __call and methods.

**Comment by tcarrio** [comment]
Definitely could do a similar check for `__call` as well.

I'm not sure where this would best be documented; for example, the behavior today of _not_ supporting magic methods isn't documented.

**Comment by tcarrio** [comment]
I opened that PR to do the checks with `isset` to support `__get` and `__isset` classes. We can discuss which path would be preferred going forward. I'd be interested in hearing more from @ramsey on it as well 🙏

**Comment by SimoTod** [comment]
Just to clarify the expected behaviour: With a fairly standard implementation (taken from the php documentation) you would not be able to access something set to null since isset is meant to return false in that case so __isset is usually (or at least often) implemented in that way. See https://onlinephp.io/c/408a9

**Comment by tcarrio** [comment]
Yes the standard behavior of `isset` will return false when a value is `null`, but when implementing the `__isset` magic method behavior it's up to the code owner.

That method signature is `public __isset(string $name): bool`, and doesn't inherently rely on `isset`, it overrides it.

At the point in the code where we've already checked for explicit properties, and explicit methods, the case for properties accessible via `__get` is not defined- and the only consistent manner to dictate that it is accessible is via `isset` and the `__isset` magic method. Otherwise, like you given an example of, it will always return _something_, even if it's a bunch of `null`s.

**Comment by SimoTod** [comment]
Not really, it's up to the code owner in the same way. Generally a magic __get will contain some ifs, return the value if those conditions are satisfied or throw if not (same as the extractor class, just a different type of error but we could catch it and let the code throw the correct exception if we are concerned about consistency). The only defect that I can see is that you have to run __get to know if it will error (and because you don't know about the implementation, you don't know what side effects it could have even if it shouldn't have any) so yeah, not ideal either and maybe the isset approach makes more sense.

However, we mentioned best practices above but then we are suggesting the code owner should implement it in a way that doesn't follow the original purpose of the method. If it's a third party library implemented in a canonical way (could be part of a dependency imported with composer), you won't have much control on it as well. https://www.php.net/manual/en/language.oop5.overloading.php

It feels like we really need a __property_exists magic (someone did try to ask for it at some point https://bugs.php.net/bug.php?id=44857 :)).
